x-pack/plugins/licensing/public/plugin.ts

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { BehaviorSubject, Subject, merge, of, timer } from 'rxjs';
+import { filter, map, pairwise, switchMap, takeUntil, tap } from 'rxjs/operators';
+import {
+  CoreSetup,
+  CoreStart,
+  Plugin as CorePlugin,
+  PluginInitializerContext,
+} from 'src/core/public';
+import { LicensingPluginSetup, ILicensingPlugin, ObjectifiedLicense } from '../common/types';
+import {
+  API_ROUTE,
+  LICENSING_SESSION,
+  LICENSING_SESSION_SIGNATURE,
+  SIGNATURE_HEADER,
+  DEFAULT_POLLING_FREQUENCY,
+} from '../common/constants';
+import { License } from '../common/license';
+import { hasLicenseInfoChanged } from '../common/has_license_info_changed';
+
+export class Plugin implements CorePlugin<LicensingPluginSetup>, ILicensingPlugin {
+  private stop$ = new Subject();
+  private removeInterceptor!: () => void;
+  private signature = '';
+  public pollingFrequency = DEFAULT_POLLING_FREQUENCY;
+  private refresher$ = new BehaviorSubject(true);
+
+  constructor(context: PluginInitializerContext) {}
+
+  private getSession(): ObjectifiedLicense | null {
+    const raw = sessionStorage.getItem(LICENSING_SESSION);
+    const signature = sessionStorage.getItem(LICENSING_SESSION_SIGNATURE);
+    const json = raw && JSON.parse(raw);
+
+    if (signature) {
+      this.signature = signature;
+    }
+
+    return json;
+  }
+
+  private setSession = (license: License) => {
+    sessionStorage.setItem(LICENSING_SESSION, JSON.stringify(license.toObject()));
+
+    if (license.signature) {
```

Comment:
the same question for `reasonUnavailable`

Reply:
These are items that need to be observable, and the `license$` observable is the best candidate since it also changes at the same time; e.g. they are co-dependent. To make this a little more palatable, I have refactored it to no longer depend on the plugin, but rather a generic signature function.

Reply:
what about 2-3 points?

Reply:
If the license changes, the signature also changes. If the signature changes, the license *sometimes* changes.

> It's not a part of public API. The only platform utilizes it internally.

Are we sure this is true? There are no legacy plugins or third-party plugins that access the signature? As long as this is true, I may be fine with moving it to be strictly be part of the plugin.

Reply:
Let's err on the side of fewer features, and add them back later if they're actually needed.